configfile: "config.yaml"

rule all:
	input:
		'chrom_sizes/dm6.chrom.sizes'

rule download_file:
	params:
		url=lambda wildcard: config['downloads'][wildcard.download_type][wildcard.sp]
	output:
		"{download_type}/{sp, dm6|hg38|mm10}.{suffix, fa.gz|gtf.gz}"
	shell:
		"wget -x -O {output} {params.url}"

rule uncompres_gz:
	input:
		"{file}.gz"
	output:
		"{file, .+\.fa}"
	shell:
		"gunzip -c {input} > {output}"

rule create_chrom_size:
	input:
		'genomes/{sp, dm6|hg38|mm10}.fa'
	output:
		'chrom_sizes/{sp}.chrom.sizes'
	shell:
		"""
		mkdir -p chrom_sizes
		samtools faidx {input}
		cut -f 1,2 {input}.fai > {output}
		"""

rule build_star_index:
	input:
		gtf= 'annotations/{sp}.gtf',
		fasta= 'genomes/{sp}.fa' 
	output:
		directory('indices/{sp}/STAR')
	shell:
		"""
		mkdir -p {output}
		STAR 	--runThreadN {threads} \
			 	--runMode genomeGenerate \
				--genomeDir {output} \
				--genomeFastaFiles {input.fasta} \
				--outFileNamePrefix {output}/ \
				--sjdbGTFfile {input.gtf}
		"""

rule create_kallisto_index:
	input:
		"transcriptomes/{sp, dm6|hg38|mm10}.fa.gz"
	output:
		"indices/{sp, dm6|hg38|mm10}/kallisto/kallisto.index"
	shell:
		"kallisto index -i {output} {input}"
