configfile: "config.yaml"

rule all:
	input:
		'indices/dm6/star',
		'genomes/dm6.fa'

rule link_index:
	params:
		p_link=lambda wildcards: config['indices'][wildcards.sp][wildcards.ix_type]
	output:
		directory('indices/{sp}/{ix_type}')
	shell:
		"""
		ln -s {params.p_link} {output}
		"""

rule link_files:
	params:
		p_link=lambda wildcard: config[wildcard.download_type][wildcard.sp]
	output:
		"{download_type}/{sp, dm6|hg38|mm10}.{suffix, fa|gtf}"
	shell:
		"""
		ln -s {params.p_link} {output}
		"""

rule create_chrom_size:
	input:
		'genomes/{sp, dm6|hg38|mm10}.fa'
	output:
		'chrom_sizes/{sp}.chrom.sizes'
	shell:
		"""
		mkdir -p chrom_sizes
		samtools faidx {input}
		cut -f 1,2 {input}.fai > {output}
		"""

rule create_kallisto_index:
	input:
		"transcriptomes/{sp, dm6|hg38|mm10}.fa.gz"
	output:
		"indices/{sp, dm6|hg38|mm10}/kallisto/kallisto.index"
	shell:
		"kallisto index -i {output} {input}"

rule create_gene_2_transcript_file:
    output:
        "annotations/gene2transcript.txt"
    script:
        "code/R/get_gene_2_transcript.R"
